### DISCUSSION TITLE
Diagnostics to disable "reportUnknownArgumentType" for lambdas only in complex scenarios

### DISCUSSION THREAD

**Discussion by Kein** [microsoft/pylance-release#2621] [Enhancement]
Consider the following:
```python
my_map = dict(numValues = {'reader': lambda x: unpack_int(x)[0]}
```

If you have `reportUnknownArgumentType` set to anything above `none` it will report this. Problem is, as far as I know, in that particular setup, you cannot annotate lambda properly with type hints. But if you disable `reportUnknownArgumentType` you will lose the ability completely. 

I don't think that's a fair trade and I think adding extra diagnostics option is lesser evil and less crumblesome than plaguing your code with `#type: ignore` all the way:
![image](https://user-images.githubusercontent.com/623756/101442770-0c715e80-392d-11eb-931e-1f5aad2dc022.png)

**Comment by erictraut** [comment]
You are correct that Python unfortunately doesn't provide any syntax for annotating the input parameters for a lambda, but you can provide an annotation outside of the lambda that provides sufficient context.

I can provide more details if you supply a working self-contained sample. The sample you've provided above has several errors and omissions.

**Comment by Kein** [comment]
The screenshot wasn't an example, it was intended to showcase how verbose and noisy it becomes.
Yes, I can provide an example:
```python
import struct
import binascii

b = b'\x01\x00\x00\x00'
unpack_int = struct.Struct('<I').unpack

my_map = dict(numValues = {'reader': lambda x: unpack_int(x)[0]})

for key, value in my_map.items():
	f = value['reader']
	value['value'] = f(b[0:4])

print(my_map)
```
I figured you can just do:
```
 #type: dict[str, dict[str, Callable[[bytelike], Union[int, bool, None]]]]
```
but the thing is, down the line I assign new keys with different value types so now I have:
```
#type: dict[str, dict[str, Union[Callable[[memoryview], Union[int, bool, None]], int, bool]]]
```
but even this wont sole the issue when I get two keys from value that are ints but they cant be summed:
![image](https://user-images.githubusercontent.com/623756/101484065-1bc2cd00-396a-11eb-8fc9-257034d3de7d.png)

 **I still insist** that the feature in question would be useful to avoid needless noise. Perhaps implemented in this way as described here:
https://github.com/microsoft/pylance-release/issues/282#issuecomment-740377743
